src/context.rs

```diff
@@ -532,11 +478,495 @@ pub fn get_version_str() -> &'static str {
+        }
+        for part in name.rsplit('\\') {
+            if part.len() > 0 {
+                name = part.to_string();
+                break;
+            }
+        }
+        let opts = sanitize_filename::Options {
+            truncate: true,
+            windows: true,
+            replacement: "",
+        };
+        let clean = sanitize_filename::sanitize_with_options(name, opts);
+        let mut iter = clean.rsplitn(2, '.');
+        let mut ext = iter.next().unwrap_or_default().to_string();
+        let mut stem = iter.next().unwrap_or_default().to_string();
+        ext.truncate(32);
+        stem.truncate(32);
+        match stem.len() {
+            0 => (ext, "".to_string()),
+            _ => (stem, format!(".{}", ext).to_lowercase()),
+        }
+    }
+}
+
+impl<'a> fmt::Display for BlobObject<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "$BLOBDIR/{}", self.name)
+    }
+}
+
+/// Errors for the [BlobObject].
+///
+/// To keep the return type small and thus the happy path fast this
+/// stores everything on the heap.
+#[derive(Debug)]
+pub struct BlobError {
+    inner: Box<BlobErrorInner>,
+}
+
+#[derive(Debug)]
+struct BlobErrorInner {
+    kind: BlobErrorKind,
+    data: BlobErrorData,
+    backtrace: failure::Backtrace,
+}
+
+/// Error kind for [BlobError].
+///
+/// Each error kind has associated data in the [BlobErrorData].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum BlobErrorKind {
+    /// Failed to create the blob.
+    CreateFailure,
+    /// Failed to write data to blob.
+    WriteFailure,
+    /// Failed to copy data to blob.
+    CopyFailure,
+    /// Blob is not in the blobdir.
+    WrongBlobdir,
+    /// Blob has a bad name.
+    ///
+    /// E.g. the name is not sanitised correctly or contains a
+    /// sub-directory.
+    WrongName,
+}
+
+/// Associated data for each [BlobError] error kind.
+///
+/// This is not stored directly on the [BlobErrorKind] so that the
+/// kind can stay trivially Copy and Eq.  It is however possible to
+/// create a [BlobError] with mismatching [BlobErrorKind] and
+/// [BlobErrorData], don't do that.
+///
+/// Any blobname stored here is the bare name, without the `$BLOBDIR`
+/// prefix.  All data is owned so that errors do not need to be tied
+/// to any lifetimes.
+#[derive(Debug)]
+enum BlobErrorData {
+    CreateFailure {
+        blobdir: PathBuf,
+        blobname: String,
+        cause: failure::Error,
+    },
+    WriteFailure {
+        blobdir: PathBuf,
+        blobname: String,
+        cause: failure::Error,
+    },
+    CopyFailure {
+        blobdir: PathBuf,
+        blobname: String,
+        src: PathBuf,
+        cause: failure::Error,
+    },
+    WrongBlobdir {
+        blobdir: PathBuf,
+        src: PathBuf,
+    },
+    WrongName {
+        blobname: PathBuf,
+    },
+}
+
+impl BlobError {
+    pub fn kind(&self) -> BlobErrorKind {
+        self.inner.kind
+    }
+
+    fn new_create_failure(
```

Comment:
`from_failure` would be a better name

Reply:
same for the others down below

Reply:
I'm not convinced here, these functions create instances with specific error kinds set, in this case `BlobErrorKind::CreateFailure`.  Using `from_failure()` suggests that the struct is made based on data from another struct, which it is not, everything is passed in as args, i don't think the `cause` counts truly as something that warrants a `from_*()`